base/util/man/man1/retrieve_dir.1

.\"
.TH retrieve_dir 1  05-December-2006  "DRMS MANPAGE" "DRMS Programmer's Manual"
.SH NAME
retrieve_dir \- DRMS retrieve a stored directory of files.
.SH SYNOPSIS
.nf
.B retrieve_dir \fR[\fB-l\fR<jsoc_flags>] \fBseries=\fR<record_set> \fBto=\fR<dest> [\fBdirkey=\fR<dirname_keyword>]


.SH DESCRIPTION
.PP
store_dir and retrieve_dir are a pair of utilities to store (and retrieve) an arbitrary 
set of files to (from) DRMS/SUMS.  retrieve_dir can easily retrieve files stored by the store_dir
utility.
.PP
Each record specified by <record_set> refers to a subdirectory within
the record's SUMS directory, <record_directory> (each record "owns" a SUMS directory).  
<dirname_keyword> names a keyword in the series, and the record's value of this keyword 
contains a directory path.  The name of the SUMS subdirectory, <subdir>, containing the files
is the directory name part of the directory path (i.e., 
it is all of the string after the last '/' in this value).  Therefore, the SUMS location
of the record's stored files is <record_directory>/<subdir>.
For each record in the <record_set>, retrieve_dir 
copies or links all files within the referenced subdirectory to the 
directory specified by the "to" argument.  Typically, the files are stored into SUMS
with the store_dir utility.  retrieve_dir will overwrite any files existing in the destination 
directory whose
name matches the name of a file to be copied from a SUMS subdirectory.  If the link flag, "-l",
is specified, then retrieve_dir will create a link to the SUMS directory instead of 
copying from the SUMS directory to the destination directory.  Keep in mind that over time
this link may become broken as SUMS may remove a stored file if it is present in SUMS 
longer than its specified retention time.
.PP
Example to store files:

  store_dir series=su_arta.TestStoreDir dirname=/auto/home2/arta/savedFilesJanuary sel=January 
note="All my January work"

Example to retrieve files:

  retrieve_dir series=su_arta.TestStoreDir to=/auto/home2/arta/restoredFilesJanuary/ dirkey=dirname

.SH FLAGS
.SS module-specific flags
.IP \fB-l\fR 20
Create a symbolic link to the SUMS directory containing the set of files of interest.
.SS jsoc-driver flags
.IP \fB-H\fR|\fB--help\fR 20
Show jsoc_main usage information.
.IP \fB-L\fR 20
Run jsoc_main driver without logging.
.IP \fB-V\fR 20
Run jsoc_main driver in verbose mode.
.IP \fB-Q\fR 20
Run jsoc_main driver in quiet mode (no terminal output).

.SH ARGUMENTS
.IP \fBseries=\fR<record_set> 20 
A series name followed by an optional record set filter 
(i.e., <name>[RecordSet_filter]).  Each record retrieved with this <record_set> query
will refer to one directory.  The contents of all such directories will be copied/linked
to <dest>.
.IP \fBto=\fR<dest> 20 
The destination directory to which the SUMS files should be copied/linked.  
Do not append a final '/' to the path (i.e., /home/arta/dir1 is acceptable, but /home/arta/dir1/
is not)
.IP \fBdirkey=\fR<dirname_keyword> 20
Name of the keyword containing the path of the directory originally
stored with store_dir.
.PP
.SH "SEE ALSO"
.BR store_dir(1)
.BR retrieve_file(1)
.BR store_file(1)



src/base/util/apps/retrieve_dir.c
/* Retrieve_dir - retrieves a dir from SUMS saved with store_dir

/* Usage:  
 * retrieve_dir {-l} series=<seriesname>[<record_spec>] to=<target dir> 
 *			{dirkey=<dirname_keyword>}
 *         
 *	Retrieve a dir from SUMS.
 *	Dir usually archived with store_dir program.
 *	The <record_spec> should resolve to one or more records.  
 *	dirkey is the index keyword in the series that has the dir name.
 *	If this series was created by the store_dir program then the 
 *	dirkey is "dirname" and this is the default.
 *	Each matched record will have the corresponding dir retrieved. 
 *	If a note is defined for the record
 *	it will be printed after the dir name.  Thus one can e.g.
 *
 *   store_dir series=su_jim.storedirs dirname=/home/xx sel=test1 note="junk"
 *
 *     will be retrieved with simply:
 *
 *  retrieve_dir su_jim.storedirs[/home/xx] to=/tmp     or
 *  retrieve_dir su_jim.storedirs[][test1] to=/tmp
 *
 *    flags:  -l make link into SUMS.
 */

#include "jsoc_main.h"
#include "drms.h"
#include "drms_names.h"
#include <unistd.h>
#include <strings.h>


/* Some important variables are pre-defined by main.  These include:
 *    DRMS_Env_t *drms_env;    DRMS session information.
 *    CmdParams_t *cmdparams;  Command line parameters.
 */
							 /*  arguments list  */
ModuleArgs_t module_args[] = { 
 {ARG_STRING, "series", "", "Series name to retrieve from"},
 {ARG_STRING, "to", "", "Target dir to cp to"},
 {ARG_STRING, "dirkey", "dirname", "Keyword containing the dir name"},
 {ARG_FLAG, "l", "0", "Link instead to the retrieved dir in the SUMS storage"}, 
 {ARG_END, NULL, NULL} /* List must end in {NULL, NULL}. */
};
					    /* Module name presented to DRMS */
char *module_name = "retrieve_dir";

/* Some global variables for this module. */
int verbose = 0;

/* Module main function. */
int DoIt(void)
{
int status = 0;

char *series;
DRMS_RecordSet_t *recordset;
int irec;
int wantlink = cmdparams_get_int(&cmdparams, "l", NULL);
//int force = cmdparams_get_int(&cmdparams, "f", NULL);
char *destination;
char *dirkey;	/* index key in the series */

series = cmdparams_get_str(&cmdparams, "series", NULL);
destination = cmdparams_get_str(&cmdparams, "to", NULL);
dirkey = cmdparams_get_str(&cmdparams, "dirkey", NULL);


recordset = drms_open_records(drms_env, series, &status);
if (status)
  {
  fprintf(stderr,"retrieve_dir target NOT_FOUND, drms_open_records failed, series=%s, status=%d.\n"
	"Aborting.\n", series,status);
  return(1);
  }
if (recordset->n > 1)
  printf("%d records matched.\n", recordset->n);

for (irec = 0; irec< recordset->n; irec++)
  {
  DRMS_Record_t *rec = recordset->records[irec];  /* pointer to current record */
  char path[DRMS_MAXPATHLEN];
  char *dirname, *cptr;

  dirname = drms_getkey_string(rec, dirkey, &status);
  drms_record_directory(rec, path, 1);
  if (*path && !status)
    { /* target dir located in SUMS */
    char cmd[DRMS_MAXPATHLEN+1024];
    cptr = rindex(dirname, '/');
    if(wantlink)
      sprintf(cmd, "ln -s %s%s %s", path, cptr, destination);
    else
      sprintf(cmd, "cp -r %s%s %s", path, cptr, destination);
    printf("cmd = %s\n", cmd); /* !!!TEMP */
    if (system(cmd))
      {
      fprintf(stderr, "retrieve_dir failed on command: %s\n",cmd);
      return(1);
      }
      char *note = drms_getkey_string(rec, "note", &status);
      printf("%s %s to %s",dirname, (wantlink ? "linked" : "retrieved"), destination);
      if (note && *note)
	printf(", note = %s\n", note);
      else
        printf(".\n");
    }
  else
    {
    printf("retrieve_dir target dir NOT_FOUND, no path or %s keyword not found in %s, abort.\n", dirkey,series);
    status = 1;
    }
  }

return(status);
}
